Animation hooks

jhsware opened the pull request:
Inferno is described as "An extremely fast, React-like JavaScript library for building modern user interfaces".  The purpose of this PR is to improve support for common UX-animations used in modern web apps such as animation on add or removing items to a list.

There is a test page implementing these animations in **examples/animations/index.html**.

This PR adds the animation hooks didAppear and willDisappear to class components. This will allow animating component when they are added or removed from the DOM. Using these hooks are optional and everything will work as before if these two hooks aren't implemented on your class component.

This PR also adds the package inferno-animation which contains a base component that can be used, extended or copied to create class components that animate. The package also contains useful helper methods to reduce the risk of bugs when creating your own customised animated components. The base component implements CSS-animations using a pattern that allows separate behaviour on appear and disappear. It also allows animating height and width of dynamically sized elements, something which isn't possible with pure CSS. The benefit of CSS-animation is that it doesn't block the main thread.

If this PR is accepted my plan is to implement similar support when moving components in the DOM-tree using a hook didMove or willMove.

As part of a major release I would suggest transferring the inferno-animation NPM-package under my control to the Inferno-project since the implementation here is very similar and migrating would be straight forward in most cases.

**IMPORTANT! This update will break existing code** that has methods named didAppear or willDisappear. As part of a major release I believe this could be acceptable. This can be avoided by adding some other marker property to indicate that the component supports animation hooks. Another solution could be to force inheriting from a known base component to activate animation hooks. However a breaking change makes the solution more in line with the rest of the API.

Closes #1238

Havunen commented:
My main concern with this PR is that, is this something inferno-js core should do? It's any way possible to do animations using Inferno js without these new hooks.

jhsware commented:
Because Inferno does element removal on unmount it is impossible to perform animations on remove without convoluted steps such as either
- cloning the element on componentWillUnmount (which is ugly, and hurts overall performance,  but works), or
- deferring removal from parent and mark object pending removal, perform the animation and then complete the removal from the parent

Both of these options add code complexity to the app with no upside. Sometimes all you want is for a button or a form error to magically animate into view. With this PR you can solve it with a simple {doShow && <AnimatedButton />}.

Other libraries such as Vue and Svelte support animations so this would put Inferno on par with them. As it stands, lacking animation support in the core is a reason not to choose Inferno over these libraries. 
https://vuejs.org/v2/guide/transitions.html
https://svelte.dev/tutorial/custom-css-transitions
The package inferno-animation is inspired by the elegant solution they implemented in Vue.

The package inferno-animation isn't a requirement for this PR to be useful, however I believe releasing it with the rest of the inferno-packages will benefit any newcomers and third party package developers by providing a de facto standard, yet keeping it opt-in.

I have tried to implement the animation hooks in such a way that it doesn't become a burden to maintain (I chose only to to implement it for class components). Obviously I will pull my weight in maintaining the inferno-animation package and do what I can with regards to the core package.

On the flip side, this is a really simple and elegant way to add common UX-animations to Inferno apps.

If you don't feel comfortable merging it into the master branch yet, an alternative could be to release it as an experimental v8 alpha. That way one could get feedback on how it works for others and I can take a look at animations when moving sibling elements and possible drag'n'drop support. I chose not to implement that in this PR because both are high risk and I want to tread lightly in the existing code base.

jhsware commented:
I want to add that this PR solves nested animations in a simple and high performing way too. This is really useful for page transitions and widgets that animate into view but also contain animated components further down their component tree.

If done outside the core rendering engine you either need to pollute the DOM with blocking class markers or pollute context with a marker to control downstream animations.

Havunen commented:
@jhsware I added add (20) and remove shuffle (20) buttons to the animation example and when you keep hitting those it fails at random. I think there is some issue when adding new items, I'm not entirely sure what is going on there, but it needs to be solved before we can merge this.

jhsware commented:
@Havunen awesome! I will look at that and also see how I can add tests to cover those cases.

jhsware commented:
@Havunen I believe I have fixed the issue. The last element in a list was removed using `dom.textContent = ''` and it was sometimes triggered earlier than other removals for some reason that I can't explain. But since all elements are to be removed anyway I just do a check that the element still exists. I am only doing this for deferred removal so it won't affect performance in normal cases.

I have added a test for this. I don't think that test is 100% robust but I can force it to fail all the times I try. Added some comments on how to force the test to fail in case someone does some digging.

Note, to make this really robust (if last element has shorter animation duration than other elements) we need to call deferComponentClassRemoval(vNode, parentDOM) instead of dom.textContent = ''. Cutting animations short will not crash the app, only cause a slight yank. I decided not to fix it right now because it affects code in several places and requires some thought.

There was also an issue resulting from the vNode.dom === null check. See commit a43500b. I am not sure what that test was trying to achieve, if it was there for a good reason let me know and I will have to investigate further.

Havunen commented on packages/inferno/src/DOM/utils/common.ts:
```diff
@@ -80,12 +90,33 @@ export function findDOMfromVNode(vNode: VNode, startEdge: boolean) {
-      removeChild(parentDOM, vNode.dom as Element);
+      // On defered removals the node might disappear because of later
+      // operations
+      if (!deferedRemoval || (parentDOM && parentDOM.contains(vNode.dom))) {
```
ugh. This will definitely hit performance .(

Havunen commented on packages/inferno/src/DOM/utils/common.ts:
```diff
@@ -80,12 +90,33 @@ export function findDOMfromVNode(vNode: VNode, startEdge: boolean) {
     if (flags & VNodeFlags.DOMRef) {
-      removeChild(parentDOM, vNode.dom as Element);
+      // On defered removals the node might disappear because of later
+      // operations
```
What does it mean: because of later operations? Did you find the node gets de-attached from the DOM?

jhsware replied:
Yes, what happens is that the last item is removed using parent.innerText = '' via clearDOM.

Because transition events can fire out of order in the browser it is unpredictable when this happens during a set of simultaneous animations (that explains the random fail). In Mocha-tests I had to provoke this by having different timeouts because setTimeout fires in a predictable order.

So on deferred removal I need to check that the element still exists. But I don't want to do that check on normal removal since it is a parentNode.contains() check that will incur a performance penalty. Added the deferredRemoval bool to indicate if this check needs to be done so it won't affect performance during synchronous removal.

jhsware replied:
And yes it could happen due to later operations too.

Havunen replied:
hmm, this routine is only called when removing direct children of given parentDOM. 

maybe we could get rid off this query selector by checking 

```
vNode.dom.parentNode === parentDOM 
```

Havunen replied:
or maybe its even enough to just check if 
```
vnode.dom.parentNode !== null
```

jhsware replied:
Smart, yeah I think that would be much better. There is a risk that the dom is moved to descendant child. This was a bad choice from me that could fail in edge cases.

I still believe that we should retain the !deferedRemoval shortcut to avoid hitting the JS/DOM-bridge for synchronous removal.

Havunen replied:
yes

jhsware replied:
I believe this can be marked as resolved after the above commit. 

Next level would be to to check that the parent isn't null and then use whatever parent to remove the node. However in that case, the dev has taken control of the DOM-element and moved it.

Havunen commented on packages/inferno-shared/index_out.js:
```diff
@@ -0,0 +1,139 @@
+(() => {
```
where did this file come from? 🤔

jhsware replied:
My bad, it sneaked in when I was trying to solve the slow builds. It was generated by esbuild and should have been deleted.

Havunen commented on packages/inferno-animation/src/utils.ts:
```diff
@@ -0,0 +1,206 @@
+}
+
+export function addClassName (node: HTMLElement, className: string) {
+  if (className) {
```
it should probably check isString(className) -- to make 0 pass?

jhsware replied:
Will investigate

Havunen commented on packages/inferno-animation/src/utils.ts:
```diff
@@ -0,0 +1,206 @@
+}
+
+export function removeClassName (node: HTMLElement, className: string) {
+  if (className) {
```
it should probably check isString(className) -- to make 0 pass?

jhsware replied:
Will investigate

Havunen commented on packages/inferno-animation/src/utils.ts:
```diff
@@ -0,0 +1,206 @@
+  var tmpDisplay = node.style.getPropertyValue('display')
+  var isDisplayNone = window.getComputedStyle(node).getPropertyValue('display') === 'none'
+  if (isDisplayNone) {
+    node.style.setProperty('display', 'block')
```
doesn't this break flex box styling?

jhsware replied:
Yes, will fix (code created when flex wasn't widely supported)

jhsware replied:
Okay, this is tricky. I am not sure we can know if it is meant to be flex or block here. This was added to support bootstrap 4 style animations in inferno-bootstrap. https://github.com/jhsware/inferno-bootstrap/blob/master/src/Modal/AnimateModal.js
I would like to keep it but add a comment about why it exists so it potentially can be removed in the future.

jhsware replied:
WON'T FIX comment in 587866ca3

Havunen commented on packages/inferno-animation/src/utils.ts:
```diff
@@ -0,0 +1,206 @@
+  if (className) {
+    const tmp = className.split(' ')
+    for (let i=0; i < tmp.length; i++) {
+      node.classList.add(tmp[i])
```
many semicolons missing in this file

jhsware replied:
Will fix

Havunen commented on packages/inferno-animation/src/utils.ts:
```diff
@@ -0,0 +1,206 @@
+  node.style.width = node.style.height = ''
+}
+
+function _getMaxTransitionDuration (/* add nodes as args*/) {
```
Shouldn't this be
```
_getMaxTransitionDuration(...nodes)
```
to avoid expensive "arguments"

jhsware replied:
Will fix

Havunen commented on packages/inferno-animation/src/utils.ts:
```diff
@@ -0,0 +1,206 @@
+    const dur = cs.getPropertyValue('transition-duration').split(',')
+    const del = cs.getPropertyValue('transition-delay').split(',')
+    const props = cs.getPropertyValue('transition-property').split(',').map((prop) => prop.trim())
+    props.forEach((prop) => {
```
this should be converted to for - loop to avoid creating unnecessary function scopes especially because its inside another loop

jhsware replied:
Will fix

Havunen commented on packages/inferno-animation/src/utils.ts:
```diff
@@ -0,0 +1,206 @@
+        }
+      }
+    })
+    const animTimeout = dur.map((v, index) => parseFloat(v) + parseFloat(del[index])).reduce((prev, curr) => prev > curr ? prev : curr, 0)
```
this could be optimized to for - loop too

jhsware replied:
Will fix

Havunen commented on packages/inferno-animation/src/utils.ts:
```diff
@@ -0,0 +1,206 @@
+    'WebkitTransition':'webkitTransitionEnd'
+  }
+
+  for(let t in transitions){
```
this probably does not need any loop at all, and because "transitions" is a constant object we could just do if - return here?

Havunen replied:
actually, it seems the whole function is constant and there is no need for the function at all, or just make it call itself and put the result into a variable.

Havunen replied:
var transitionEndName = (function whichTransitionEvent() { ... })()

jhsware replied:
Will investigate

jhsware replied:
FIXED in 695fe367e

Havunen commented on packages/inferno-animation/src/utils.ts:
```diff
@@ -0,0 +1,206 @@
+ * will be animations that fail to complete before the timeout is triggered.
+ */
+export function registerTransitionListener(nodes: HTMLElement[], callback: Function, noTimeout: boolean = false) {
+  // I am doing this lazily because there where issues with document being undefined
```
const hasDocumentAvailable: boolean = typeof document !== 'undefined';

jhsware replied:
Will investigate

jhsware replied:
FIXED in 695fe367e

Havunen commented on packages/inferno-animation/src/utils.ts:
```diff
@@ -0,0 +1,206 @@
+
+  // Fallback if transitionend fails
+  // This is disabled during debug so we can set breakpoints
+  if (!window.__DEBUG_ANIMATIONS__ && !noTimeout) {
```
this debug thing could be moved to its own function and then prevent calling that funciton using: 
```
if (process.env.NODE_ENV !== 'production') {
}

```

to keep it inline with other infernojs debug messages

jhsware replied:
@Havunen QUESTION: This would make debugging animations in production code using slowed down animations or by setting a breakpoint impossible. Do we want this? Should we provide an alternative solution to allow enabling this?

Havunen replied:
using process.env.NODE_ENV production check would strip off this code from the production build. IMO debugging is for development bundles, at least this is how it is with all other code in infernojs

jhsware replied:
FIXED in 538601bb0

Havunen commented on packages/inferno/src/DOM/mounting.ts:
```diff
@@ -152,7 +159,16 @@ function createClassMountCallback(instance) {
 
-export function mountClassComponentCallbacks(ref, instance, lifecycle: Function[]) {
+function addAppearAnimationHook(animations: AnimationQueues, instance) {
+  // Only do animations in browser
```
we can get rid of this "if" statement from runtime by checking const addAppearAnimationHook =  isDocumentAvailable ? emptyFn : function () { ... }

jhsware replied:
Will investigate

jhsware replied:
@Havunen so it appears that the tests will fail if we check for window on loading. Is there some other way you know of that can perform the check later and provide the same optimisation? Is it accessing the window object that is slow? Should we assign a `let addAppearAnimationHook` on first call?

Havunen replied:
infernojs is already doing it: 

https://github.com/infernojs/inferno/blob/bcfcfd8fd786681ab9cb666a7deb39738b34d206/packages/inferno/src/DOM/rendering.ts#L10-L18

jhsware replied:
@Havunen after more investigation and the fact that I couldn't cause an SSR test for animation hooks to fail I realised that the SSR-rendering is done via an entirely different code path that doesn't trigger animation hooks anyway. So I simply remove the if statement. Added test in inferno-server to catch this if SSR-rendering changes in the future.

FIXED in e2c8e0b08

Havunen commented on packages/inferno/src/DOM/unmounting.ts:
```diff
@@ -35,53 +35,82 @@ export function unmount(vNode) {
+  if (animations.willDisappear.length > 0) {
+    // Wait until animations are finished before removing actual dom nodes
+    // Be aware that the element could be removed by a later operation
+    callAllAnimationHooks(animations.willDisappear, () => { if (dom) dom.textContent = '' });
```
function creation should be moved outside runtime of "clearDOM"

jhsware replied:
Will fix

jhsware replied:
@Havunen so I solved this by looking at how we optimised removeVNodeDOM. I don't fully grasp why this is better, but I get that it is a different way to create the anonymous callback function.

In other words, don't assume I know what I am doing and if I have misunderstood what you were going for, also take a look at `function removeVNodeDOM(...)` in common.ts. It uses the same pattern.

FIXED in 7cf58ef09

Havunen commented on packages/inferno/src/DOM/unmounting.ts:
```diff
@@ -35,53 +35,82 @@ export function unmount(vNode) {
   }
 }
+
+function addDisappearAnimationHook(animations: AnimationQueues, instance, dom: Element) {
```
same comment about hoisting window !== 'undefined'

jhsware replied:
Will fix

jhsware replied:
FIXED in e2c8e0b08

Havunen commented on packages/inferno/src/DOM/utils/common.ts:
```diff
@@ -80,12 +90,33 @@ export function findDOMfromVNode(vNode: VNode, startEdge: boolean) {
+  do {
+    const fn = animationQueue.pop();
+    // This check shouldn't be needed but TS complains so adding it
+    if (fn !== undefined) {
```
its needed because last pop will give empty, so the loop here is little bit wierd... you could do while(pop() ...)   or then just regular for loop if we need to calculate animations left

Havunen replied:
is there any reason to actally "pop" from the array cant we just do linear - read ? simple for loop ?

Havunen replied:
the issue with the latest loop is that its reading length of the array and the same array is mutated using pop so length does not get cached. 

We can work around by not reading length.
shift for first to last and pop for last to first
```js
 let listener;
  while ((callback = array.shift()) !== undefined) {
    callback();
  }
```
but regular for loop will be way faster if we do not need to "empty" the array

Havunen replied:
or, if we know that there cannot be new items added to the array during the loop, but we need to empty it then we can do regular for loop and then at the end of the loop, clear it by reference  

array.length = 0;

jhsware replied:
@Havunen the reason for .pop(): when multiple removes are triggered in the same patch you get multiple calls to callAllAnimationHooks. I tried a for loop but it failed during tests and no elements where removed. I couldn't really figure out what was happening because I can't step through the code properly (misaligned cursor during debugging). 

Implemented you suggested optimisation of the while loop.

FIXED in ca9b1277a

Havunen commented on packages/inferno/src/DOM/utils/common.ts:
```diff
@@ -80,12 +90,33 @@ export function findDOMfromVNode(vNode: VNode, startEdge: boolean) {
+  } while (animationQueue.length > 0);
+}
+
+function clearVNodeDOM(vNode: VNode, parentDOM: Element, deferedRemoval = false) {
```
there should be no optional parameters in inferno js code because that goes in V8 engine trampoline routine which is slow

jhsware replied:
Will fix

Havunen commented on packages/inferno-animation/src/AnimatedComponent.ts:
```diff
@@ -0,0 +1,96 @@
+    }, false)
+  
+    // 4. Activate target state
+    setTimeout(() => {
```
what is the reason for timeout and why 5ms?

jhsware replied:
@Havunen the setTimeout is needed otherwise the browser engine optimises the following sequence of operations:

```
// 2. Activate transition
addClassName(dom, animCls.active)
// 4. Activate target state
setDimensions(dom, width, height)
removeClassName(dom, animCls.start)
addClassName(dom, animCls.end)
```

and immediately goes to the end state without a transition. Setting it too short also triggers optimisation. I could switch to requestAnimationFrame, I believe the choice of setTimeout was to support IE9 without polyfill a couple of years back.

I will switch to requestAnimationFrame unless you have another opinion.

Havunen replied:
Yeah, IE9 and IE10 are not officially supported by inferno js and I personally hope all users stop using them.

jhsware replied:
FIXED in aeebd7240

Havunen commented:
I just tested the example again and its still failing when clicking "add and shuffle 20" and "20 do 5 shuffle" -buttons 

![image](https://user-images.githubusercontent.com/2021355/112515089-a22adf80-8d9e-11eb-8ee8-0ff3598feeec.png)

jhsware commented:
@Havunen Regarding the DOMException, I am not able to reproduce your issue when manually clicking all the the buttons intensely in semi-random order and multiple times before animations complete.

Are you able to do this consistently with the head of the branch after an `npm run build` in root? What steps do I need to perform to get the same result?

I have tried both in Chrome and Safari on Mac.

jhsware commented on packages/inferno-animation/src/utils.ts:
```diff
@@ -0,0 +1,206 @@
+     * Perform cleanup
+     */
+    rootNode.removeEventListener(transitionEndName, onTransitionEnd, false)
+    callback && callback()
```
Will fix

Havunen commented on packages/inferno-animation/src/AnimatedComponent.ts:
```diff
@@ -83,14 +83,14 @@ export default class AnimatedComponent extends Component<Props> {
     // Needs to be done after activating so timeout is calculated correctly
     registerTransitionListener(dom, function () {
```
Can we directly set  registerTransitionListener(dom, callback, false) here to avoid creating function context?

jhsware replied:
Will investigate

jhsware replied:
@Havunen from what I see we have three alternatives:

1. current solution using closure
2. using .bind on a component method and creating an object that contains dom and animCls to pass as this
3. setting dom and animCls on the component and accessing it from method that is added to the component

Alt. 2 I don't think has any performance benefit
Alt. 3 pollutes the component and could cause unexpected bugs

Both alt. 2 & 3 reduce code readability quite a lot, so I am not sure if it is worth it given that animations are expensive operations anyway.

How do you feel about it?

Havunen replied:
ok

jhsware commented:
@Havunen UPDATE! I managed to get the same DOMException as you once, but I can't repeat it and I believe I reloaded the window to retry thus not allowing me to check if any element was left behind.

Is there a case where VNodes can be reused after being unmounted thus potentially existing in two places at once? (both in the animation queue and the virtual DOM)

Havunen commented:
Yeah, it's easy to reproduce, just click those add 20 or remove 20 buttons, they should always even out. Adding 20 - removing 20 should result in 0 items in the list.

video: https://streamable.com/vq8i0h

> Is there a case where VNodes can be reused after being unmounted thus potentially existing in two places at once? (both in the animation queue and the virtual DOM)

nope

jhsware commented:
@Havunen excellent! Thanks for the screencast, now I can reproduce it consistently. I'll investigate and figure out what is happening. I literally never did it in that order. I am pretty sure it is the `dom.textContent = ''` that is the culprit, but I will need to investigate.

jhsware commented:
@Havunen The issue with added elements being removed when the last item is removed is FIXED by a376f51fe

I feel the code is in the right place and does what it is supposed to do, which is handles clearDOM calls differently when we have running animations. Feel free to disagree :D

There is still an obvious case it doesn't handle. If a later mutation without animations calls clearDOM it will clear immediately. To solve that we would need to keep a list of parent DOM elements that are currently animating and perform "slow" deletes if we get calls that match that list. Obviously that would slow down the clearDOM operation during animation and I am not sure it is worth it given that the only effect will be that the animations are cut short. In most practical use cases such situations won't be an issue.

restlessmedia commented:
This has been a fascinating code review!

Havunen commented:
> The issue with added elements being removed when the last item is removed is FIXED by a376f51

Yes, makes sense 👍

Havunen commented:
> There is still an obvious case it doesn't handle. If a later mutation without animations calls clearDOM it will clear immediately. To solve that we would need to keep a list of parent DOM elements that are currently animating and perform "slow" deletes if we get calls that match that list. Obviously that would slow down the clearDOM operation during animation and I am not sure it is worth it given that the only effect will be that the animations are cut short. In most practical use cases such situations won't be an issue.

I believe this can be fixed by checking if parentNode is not null before trying to remove the node. Inferno will not move node from parent node to another ( at least not yet ) so that check should be enough.

It would be good to have test case for this. If we ever try to implement optimization for static vNode trees then this might become an issue

jhsware commented:
@Havunen I was thinking more along the lines that when the last element is to be removed synchronously it clears the entire content of the parent. If a delete animation is still running it will be removed in that same operation. We are talking a very weird way to structure the components of the app so that case feels more academic than a real issue. 

I'll investigate your suggestion and see if something needs to be added.

On a side note, the build times had quite a big impact on productivity. I estimate that I have spent in total 2h+ Today just waiting for the project to build so I can test every individual commit in isolation.

Havunen commented:
I just re-run krauset/js-framework-benchmark and ui benchmark with and without the changes in this PR and everything seems to be within the error margin of the results. So Whenever ready I think it's good to be merged.

Havunen commented on packages/inferno-animation/src/utils.ts:
```diff
@@ -213,7 +213,7 @@ export function registerTransitionListener(nodes: HTMLElement[], callback: Funct
   // Fallback if transitionend fails
   // This is disabled during debug so we can set breakpoints
-  if (!window.__DEBUG_ANIMATIONS__ && !noTimeout) {
+  if (!isDebugAnimationsSet() && !noTimeout) {
```
I meant moving the contents of this function outside, registerTransitionListener  otherwise we create 2 additional function scopes even if noTimeout or debug is false.

Also process.env.NODE_ENV !== 'production' check needs to be inlined here, because its special compiler time check, which will be replaced with true or false depending on the build target, then it will be fully removed from the output as a result

jhsware replied:
Now I have checked that the actual code is removed in minified production build (i.e. the string '__INFERNO_ANIMATION_DEBUG__' doesn't exist). I also changed the name of the variable you set to window.__INFERNO_ANIMATION_DEBUG__.

In dev builds you can set `window.__INFERNO_ANIMATION_DEBUG__ = true` in the console to start debugging if an animation isn't behaving as expected (usually meaning you have forgotten to include the CSS, but could be caused by other issues if you have created your own animation component from scratch).

FIXED in 4366159fe

Havunen commented:
Was there anything still left to do? or shall we merge this?

jhsware commented:
@Havunen I don't believe we have anything left to do at this point so I think it can be merged.

Next step would be for me to write documentation, but I would like to discuss how and where it would be appropriate.

I would also like to discuss a feature roadmap for inferno-animation and animation hooks that I could investigate going forward. I have some ideas.

Finally I would like to transfer the npm-package inferno-animation to the project if you agree to that. It would however require a major release of Inferno because I synched major version numbers.

I'll move these questions to Slack so no need to answer anything here. Just leaving it for posterity... :D
